src/main/java/org/testng/internal/Parameters.java

```diff
@@ -264,35 +264,81 @@ else if (ITestResult.class.equals(c)) {
                                        MethodParameters params,
                                        XmlSuite xmlSuite) {
     List<Object> vResult = Lists.newArrayList();
-    for (int i = 0, j = 0; i < parameterTypes.length; i++) {
-      if (j < parameterNames.length) {
-        String p = parameterNames[j];
-        String value = params.xmlParameters.get(p);
-        if (null == value) {
-          // try SysEnv entries
-          value = System.getProperty(p);
+    if (optionalValues.length != parameterNames.length) {
+      FilterOutInJectedTypesResult filterOutResult = filterOutInJectedTypesFromOptionalValues(parameterTypes, optionalValues);
+      optionalValues = filterOutResult.getOptionalValues();
+      parameterTypes = filterOutResult.getParameterTypes();
+    }
+    for (int i = 0; i < parameterNames.length; i++) {
+      String p = parameterNames[i];
+      String value = params.xmlParameters.get(p);
+      if (null == value) {
+        // try SysEnv entries
+        value = System.getProperty(p);
+      }
+      if (null == value) {
+        if (optionalValues != null) {
+            value = optionalValues[i];
         }
         if (null == value) {
-          if (optionalValues != null) {
-            value = optionalValues[i];
-          }
-          if (null == value) {
             throw new TestNGException("Parameter '" + p + "' is required by "
                     + methodAnnotation
                     + " on " + prefix + " "
                     + name
                     + " but has not been marked @Optional or defined\n"
                     + (xmlSuite.getFileName() != null ? "in "
-                    + xmlSuite.getFileName() : ""));
-          }
+                            + xmlSuite.getFileName() : ""));
+        }
+    }
+      vResult.add(convertType(parameterTypes[i], value, p));
+    }
+    return vResult;
+  }
+  
+  /**
+   * Remove injected types from parameterTypes and optionalValues
+   * @param parameterTypes
+   * @param optionalValues
+   * @return FilterOutInJectedTypesResult
+   */
+  static FilterOutInJectedTypesResult filterOutInJectedTypesFromOptionalValues(Class<?>[] parameterTypes, String[] optionalValues) {
+    List<Class<?>> typeList = Lists.newArrayList();
+    List<String> optionalValueList = Lists.newArrayList();
+    Collections.addAll(typeList, parameterTypes);
```

Comment:
You can add the `Lists.<List<T>>newArrayList(T[])` helper method.

Reply:
Oh, you remind me. Will change it.